internal/retryafter_test.go

```diff
@@ -0,0 +1,100 @@
+package internal
+
+import (
+	"github.com/stretchr/testify/assert"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"testing"
+	"time"
+)
+
+func response503() *http.Response {
+	return &http.Response{
+		StatusCode: http.StatusServiceUnavailable,
+		Header:     map[string][]string{},
+	}
+}
+
+func TestExtractRetryAfterHeaderDelaySeconds(t *testing.T) {
+	// Generate random n > 0 int
+	rand.Seed(time.Now().UnixNano())
+	retryIntervalSec := rand.Intn(9999)
+
+	// Generate a 503 status code response with Retry-After = n header
+	resp := response503()
+	resp.Header.Add(retryAfterHTTPHeader, strconv.Itoa(retryIntervalSec))
+
+	expectedDuration := OptionalDuration{
+		Defined:  true,
+		Duration: time.Second * time.Duration(retryIntervalSec),
+	}
+	assert.Equal(t, expectedDuration, ExtractRetryAfterHeader(resp))
+}
+
+func TestExtractRetryAfterHeader429StatusCode(t *testing.T) {
+	// Generate random n > 0 int
+	rand.Seed(time.Now().UnixNano())
+	retryIntervalSec := rand.Intn(9999)
+
+	// Generate a 429 status code response with Retry-After = n header
+	resp := response503()
+	resp.StatusCode = http.StatusTooManyRequests
+	resp.Header.Add(retryAfterHTTPHeader, strconv.Itoa(retryIntervalSec))
+
+	expectedDuration := OptionalDuration{
+		Defined:  true,
+		Duration: time.Second * time.Duration(retryIntervalSec),
+	}
+	assert.Equal(t, expectedDuration, ExtractRetryAfterHeader(resp))
+}
```

Comment:
These 2 tests are almost identical. It would be nice to make them one test with 2 sub-tests. It will eliminate a lot of lines.